pkg/config/database.go

```diff
@@ -35,30 +38,74 @@ func (d *Database) Open(logger *logging.Logger) (*icingadb.DB, error) {
 		driver.Register(logger)
 	})
 
-	config := mysql.NewConfig()
+	var dsn string
+	switch d.Type {
+	case "mysql":
+		config := mysql.NewConfig()
 
-	config.User = d.User
-	config.Passwd = d.Password
-	config.Net = "tcp"
-	config.Addr = net.JoinHostPort(d.Host, fmt.Sprint(d.Port))
-	config.DBName = d.Database
-	config.Timeout = time.Minute
+		config.User = d.User
+		config.Passwd = d.Password
+		config.Net = "tcp"
+		config.Addr = net.JoinHostPort(d.Host, fmt.Sprint(d.Port))
+		config.DBName = d.Database
+		config.Timeout = time.Minute
+		config.Params = map[string]string{"sql_mode": "ANSI_QUOTES"}
 
-	tlsConfig, err := d.TlsOptions.MakeConfig(config.Addr)
-	if err != nil {
-		return nil, err
-	}
+		tlsConfig, err := d.TlsOptions.MakeConfig(config.Addr)
+		if err != nil {
+			return nil, err
+		}
 
-	if tlsConfig != nil {
-		config.TLSConfig = "icingadb"
-		if err := mysql.RegisterTLSConfig(config.TLSConfig, tlsConfig); err != nil {
-			return nil, errors.Wrap(err, "can't register TLS config")
+		if tlsConfig != nil {
+			config.TLSConfig = "icingadb"
+			if err := mysql.RegisterTLSConfig(config.TLSConfig, tlsConfig); err != nil {
+				return nil, errors.Wrap(err, "can't register TLS config")
+			}
+		}
+
+		dsn = config.FormatDSN()
+	case "pgsql":
+		uri := &url.URL{
+			Scheme: "postgres",
+			User:   url.UserPassword(d.User, d.Password),
+			Host:   net.JoinHostPort(d.Host, strconv.FormatInt(int64(d.Port), 10)),
+			Path:   "/" + url.PathEscape(d.Database),
+		}
+
+		if _, err := d.TlsOptions.MakeConfig(uri.Host); err != nil {
+			return nil, err
 		}
-	}
 
-	dsn := config.FormatDSN()
+		query := url.Values{"connect_timeout": {"60"}, "binary_parameters": {"yes"}}
+		if d.TlsOptions.Enable {
+			if d.TlsOptions.Insecure {
+				query["sslmode"] = []string{"require"}
+			} else {
+				query["sslmode"] = []string{"verify-full"}
+			}
 
-	db, err := sqlx.Open("icingadb-mysql", dsn)
+			if d.TlsOptions.Cert != "" {
+				query["sslcert"] = []string{d.TlsOptions.Cert}
+			}
+
+			if d.TlsOptions.Key != "" {
+				query["sslkey"] = []string{d.TlsOptions.Key}
+			}
+
+			if d.TlsOptions.Ca != "" {
+				query["sslrootcert"] = []string{d.TlsOptions.Ca}
+			}
+		} else {
+			query["sslmode"] = []string{"disable"}
+		}
+
+		uri.RawQuery = query.Encode()
+		dsn = uri.String()
+	default:
+		return nil, errors.Errorf(`unknown database type %q, must be one of: "mysql", "pgsql"`, d.Type)
```

Comment:
I would use a function for both error constructions as with https://github.com/Icinga/icingadb/blob/master/pkg/logging/logging.go#L129

Reply:
For finding usages?

Reply:
And for just a single place to change code whenever we need to change something.